### ISSUE TITLE
BearSSL?

### ISSUE THREAD

**Issue by markrad** [LiamBindle/MQTT-C#79]
This not an issue.

For a proof of concept I needed to find an alternative library to Paho MQTT and a different TLS library in order to demonstrate how one might replace those components in a project that I am associated with. I looked through my options in GitHub and I liked your particular implementation. It was clean and simple and provided a mechanism that allowed me to use an alternative TLS library.

BearSSL (https://bearssl.org/) is a good choice for small systems since it does not allocate any memory which is what I was looking for. As far as I can tell your MQTT library also doesn't allocate any memory which made it an ideal partner. 

The upshot of this is that I now have a PAL layer for your library that can use BearSSL as its TLS layer. Currently the code is fairly messy but, if you are interested in a pull request, I'm happy to clean it up and send it to you. If not, then I'll just use it as it is, debug code and all. 

The cmake file is pretty hacky and I'm not much of a cmake expert so I may leave that to you. It works but I don't know how to write a find package (and didn't have the time to spend on it) so it just looks for the header and the library.

Anyway, let me know. Nice job BTW.

**Comment by LiamBindle** [comment]
@markrad Yeah, MQTT-C sounds like a good fit for your project. Also, you're correct in that MQTT-C doesn't allocate any memory.

I haven't seen BearSSL before, but it looks cool and seems like a good fit for MQTT-C. I'd be happy to add it to the default available PALs. I think you understand the spirit of MQTT-C (i.e. simple and hackable), but perhaps I could suggest a few things:
1. The defaults PALs are intended to be examples as much as anything else, so please keep the PAL implementation simple, obvious, and clean.
2. No dynamic memory allocation
3. Beside the PAL, please include (1) at least a publisher example (e.g. examples/bearssl_publisher.c) and (2) a template for opening BearSSL sockets (examples/templates/bearssl_sockets.h)

No worries about the CMake files. Honestly, my preference would be to omit the build files and just include the PAL, examples, and template. Since MQTT-C is only 2 files, in practice I think it's a lot easier if users just build MQTT-C into their own build systems rather than try to interface builds.

Looking forward to it!
-Liam

**Comment by markrad** [comment]
I haven't actually forgotten about this. I needed to ensure that I could disconnect and reconnect because my use case has a TTL for the password which means I need to reauthenticate before that TTL expires. MQTT 3.1 does not offer in session reauthentication so I needed to disconnect and reconnect. I beat my head against a bug. When I reconnected it would negotiate the TLS but it would get a TLS encrypted alert as soon as I attempted to send any data. Anyway it was a face palm moment. I called mqtt_disconnect but I didn't call mqtt_sync afterwards so after reconnecting the very first thing your code did (correctly) was send the DISCONNECT packet that was still queued. Not sure why that would cause an encrypted alert but such is life. I'm cleaning up the code now for a pull request.
